### PR TITLE
Feature/openmpi two level namespace

### PR DESCRIPTION
**Pull request by climbfuji**
This PR does the following:
- Add option to compile `openmpi` without flat namespace (aka two-level namespaces) - corresponding flags borrowed from `mpich`
- Rename `nceplibs-bundle` to `nceplibs-bundle-env`
- Add new package `py-pyhdf`, add it to `jedi-base-env`
- Bug fix for building `hdf` on Ubuntu 20.04 aarch64 - see also https://github.com/spack/spack/pull/30522
- Get timeout for web requests with `urllib` from spack config, same as for `curl` (see corresponding PR for the authoritative spack repo https://github.com/spack/spack/pull/30468)

See https://github.com/NOAA-EMC/spack-stack/pull/131 for the corresponding changes in `spack-stack` and for testing.

### CODE REVIEW

File: var/spack/repos/builtin/packages/openmpi/package.py [hunk 1]
```diff
@@ -713,6 +722,8 @@ def configure_args(self):
         # Hwloc support
         if '~internal-hwloc' in spec and spec.satisfies('@1.5.2:'):
             config_args.append('--with-hwloc={0}'.format(spec['hwloc'].prefix))
+        elif '+internal-hwloc' and spec.satisfies('@4.1.3:'):
```
Comment (climbfuji) [inline var/spack/repos/builtin/packages/openmpi/package.py:725]: Note: `--with-hwloc=internal` seems to be no longer the default in `openmpi@4.1.3`.

File: var/spack/repos/builtin/packages/openmpi/package.py [hunk 2]
```diff
@@ -796,11 +807,22 @@ def configure_args(self):
                 config_args.append('--disable-wrapper-runpath')
 
             # Add extra_rpaths and implicit_rpaths into the wrappers.
-            wrapper_ldflags.extend([
-                self.compiler.cc_rpath_arg + path
-                for path in itertools.chain(
-                    self.compiler.extra_rpaths,
-                    self.compiler.implicit_rpaths())])
+            if wrapper_ldflags and any(self.compiler.linker_arg \
```
Comment (climbfuji) [inline var/spack/repos/builtin/packages/openmpi/package.py:810]: This complicated logic is needed to combine multiple `-Wl,-rpath` configuration options in the correct way.

File: var/spack/repos/builtin/packages/openmpi/package.py [hunk 3]
```diff
@@ -815,10 +837,26 @@ def configure_args(self):
...
+
         if wrapper_ldflags:
             config_args.append(
                 '--with-wrapper-ldflags={0}'.format(' '.join(wrapper_ldflags)))
 
+        # Namespaces (macOS only) - part 2
+        if '+two_level_namespace' in spec and spec.satisfies('platform=darwin'):
+            config_args.append('LIBS={}-commons,use_dylibs'.format(
+                self.compiler.linker_arg))
+        elif spec.satisfies('platform=darwin'):
```
Comment (climbfuji) [inline var/spack/repos/builtin/packages/openmpi/package.py:856]: Note to self. Need to retest with flat namespace variant.

Comment (climbfuji) [reply var/spack/repos/builtin/packages/openmpi/package.py:856]: Done. Still works.

File: var/spack/repos/builtin/packages/py-pyhdf/package.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    depends_on('python@3.2:', type=('build', 'run'))
+
+    # Dependencies
+    depends_on('zlib', type=('build', 'run'))
+    depends_on('hdf', type=('build', 'run'))
+    depends_on('py-numpy', type=('build', 'run'))
+    depends_on('jpeg', type=('build', 'run'))
+
+    def setup_build_environment(self, env):
+        # DH* There should be direct attributes to use
```
Comment (climbfuji) [inline var/spack/repos/builtin/packages/py-pyhdf/package.py:32]: Will be done in a follow-up PR, see https://github.com/NOAA-EMC/spack/issues/59

**Comment by climbfuji** [comment]
@srherbener FYI

**Review by kgerheiser — APPROVED** [review]

**Pull request merged**